server/blob/blob-objectstorage/src/main/java/org/apache/james/blob/objectstorage/ObjectStorageBlobStore.java

```diff
@@ -101,17 +106,26 @@ public void close() {
         ObjectStorageBucketName resolvedBucketName = bucketNameResolver.resolve(bucketName);
 
         return Mono.fromCallable(() -> blobIdFactory.forPayload(data))
-            .flatMap(blobId -> {
-                Payload payload = payloadCodec.write(data);
+            .flatMap(blobId -> blobExistenceTester.exists(resolvedBucketName, blobId)
+                .flatMap(exists -> {
+                    if (exists) {
+                        return Mono.just(blobId);
+                    }
+                    return save(resolvedBucketName, blobId, data);
```

Comment:
is there any case where the blob could have been deleted just after the `exists` call that would lead to a state where the blob is no longer in the blobstore?

Reply:
At the moment the operation started it did not exist so I consider this behaviour valid.

Reply:
which operation?
let's see with several threads:
T0 -> exists == false
T0 -> save
T1 -> delete the blob from the blobstore
T2 -> exists == true
T1 -> delete the blobId from cassandra
T2 -> doesn't save the blob in blobstore